File: .claude/bmad-template/lints/typescript/check-file-sizes.js
#!/usr/bin/env node
/**
 * Check that source files stay under a configurable line limit.
 *
 * Large files indicate mixed responsibilities. They are harder for agents
 * to reason about and more likely to cause merge conflicts.
 *
 * Usage:
 *   node lints/typescript/check-file-sizes.js [--max-lines N] [--path DIR]
 *
 * Exit codes:
 *   0 — No violations
 *   1 — One or more files exceed the limit
 */

const fs = require("fs");
const path = require("path");

const DEFAULT_MAX_LINES = 300;
const DEFAULT_PATH = ".";
const EXTENSIONS = new Set([".ts", ".tsx", ".js", ".jsx"]);
const SKIP_DIRS = new Set([
  "node_modules", ".git", "dist", "build", ".next", "coverage",
  "__pycache__", "venv", ".venv",
]);

function findSourceFiles(root) {
  const files = [];

  function walk(dir) {
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (SKIP_DIRS.has(entry.name)) continue;

      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.isFile() && EXTENSIONS.has(path.extname(entry.name))) {
        files.push(fullPath);
      }
    }
  }

  walk(root);
  return files.sort();
}

function checkFileSize(filepath, maxLines) {
  let content;
  try {
    content = fs.readFileSync(filepath, "utf-8");
  } catch {
    return null;
  }

  const lineCount = content.split("\n").length;
  if (lineCount <= maxLines) return null;

  const stem = path.basename(filepath, path.extname(filepath));
  return [
    `VIOLATION: File Size Limit`,
    `FILE: ${filepath}:${lineCount}`,
    `RULE: Source files must be under ${maxLines} lines`,
    `WHY: Large files indicate mixed responsibilities. They are harder for agents to reason about and more likely to cause merge conflicts.`,
    `FIX: Extract related functions into a new module. Split by responsibility (e.g., ${stem}.core.ts, ${stem}.helpers.ts).`,
    "",
  ].join("\n");
}

function main() {
  const args = process.argv.slice(2);
  let maxLines = DEFAULT_MAX_LINES;
  let rootPath = DEFAULT_PATH;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--max-lines" && args[i + 1]) {
      const parsed = parseInt(args[i + 1], 10);
      if (!Number.isFinite(parsed) || parsed <= 0) {
        console.error(`Error: --max-lines requires a positive integer, got "${args[i + 1]}"`);
        process.exit(1);
      }
      maxLines = parsed;
      i++;
    } else if (args[i] === "--path" && args[i + 1]) {
      rootPath = args[i + 1];
      i++;
    }
  }

  if (!fs.existsSync(rootPath) || !fs.statSync(rootPath).isDirectory()) {
    console.error(`Error: ${rootPath} is not a directory`);
    process.exit(1);
  }

  const files = findSourceFiles(rootPath);
  const violations = [];

  for (const f of files) {
    const violation = checkFileSize(f, maxLines);
    if (violation) violations.push(violation);
  }

  if (violations.length === 0) {
    console.log(`OK: All ${files.length} files are under ${maxLines} lines`);
    process.exit(0);
  }

  for (const v of violations) {
    console.log(v);
  }

  console.log(`FAILED: ${violations.length} file(s) exceed ${maxLines} lines`);
  process.exit(1);
}

main();

File: .claude/bmad-template/lints/python/check_file_sizes.py
#!/usr/bin/env python3
"""Check that source files stay under a configurable line limit.

Large files indicate mixed responsibilities. They are harder for agents
to reason about and more likely to cause merge conflicts.

Usage:
    python lints/python/check_file_sizes.py [--max-lines N] [--path DIR]

Exit codes:
    0 — No violations
    1 — One or more files exceed the limit
"""

import argparse
import sys
from pathlib import Path

DEFAULT_MAX_LINES = 300
DEFAULT_PATH = "."
EXTENSIONS = {".py"}
SKIP_DIRS = {"__pycache__", ".git", "node_modules", "venv", ".venv", ".tox", "dist", "build"}


def find_source_files(root: Path) -> list[Path]:
    """Recursively find source files, skipping excluded directories."""
    files = []
    for item in root.rglob("*"):
        if any(part in SKIP_DIRS for part in item.parts):
            continue
        if item.is_file() and item.suffix in EXTENSIONS:
            files.append(item)
    return sorted(files)


def check_file_size(filepath: Path, max_lines: int) -> str | None:
    """Return a violation string if the file exceeds max_lines, else None."""
    try:
        lines = filepath.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return None

    count = len(lines)
    if count <= max_lines:
        return None

    return (
        f"VIOLATION: File Size Limit\n"
        f"FILE: {filepath}:{count}\n"
        f"RULE: Source files must be under {max_lines} lines\n"
        f"WHY: Large files indicate mixed responsibilities. They are harder for "
        f"agents to reason about and more likely to cause merge conflicts.\n"
        f"FIX: Extract related functions into a new module. Split by responsibility "
        f"(e.g., {filepath.stem}_core.py, {filepath.stem}_helpers.py).\n"
    )


def main() -> int:
    parser = argparse.ArgumentParser(description="Check source file sizes")
    parser.add_argument("--max-lines", type=int, default=DEFAULT_MAX_LINES,
                        help=f"Maximum lines per file (default: {DEFAULT_MAX_LINES})")
    parser.add_argument("--path", type=str, default=DEFAULT_PATH,
                        help=f"Root directory to scan (default: {DEFAULT_PATH})")
    args = parser.parse_args()

    root = Path(args.path)
    if not root.is_dir():
        print(f"Error: {root} is not a directory", file=sys.stderr)
        return 1

    files = find_source_files(root)
    violations = []

    for f in files:
        violation = check_file_size(f, args.max_lines)
        if violation:
            violations.append(violation)

    if not violations:
        print(f"OK: All {len(files)} files are under {args.max_lines} lines")
        return 0

    for v in violations:
        print(v)

    print(f"FAILED: {len(violations)} file(s) exceed {args.max_lines} lines")
    return 1


if __name__ == "__main__":
    sys.exit(main())
